diff --git a/amazons3_cors.module b/amazons3_cors.module
index 36644b1..d75d054 100644
--- a/amazons3_cors.module
+++ b/amazons3_cors.module
@@ -363,6 +363,18 @@ function amazons3_cors_upload_value(&$element, $input = FALSE, $form_state = NUL
     $file->status = 0;
     $file->timestamp = REQUEST_TIME;
 
+    // Make sure the file destination is not already in the file_managed table.
+    $existing_files = file_load_multiple(array(), array('uri' => $file->uri));
+    if (!empty($existing_files)) {
+      $safe_uri = file_create_filename($file->filename, 's3://' . $base_dir);
+      if ($file->uri != $safe_uri) {
+        $final_uri = file_unmanaged_copy($file->uri, $safe_uri);
+        if ($final_uri) {
+          $file->uri = $final_uri;
+        }
+      }
+    }
+
     // Save the uploaded file to the file_managed table.
     if ($file = _amazons3_cors_file_save($file)) {
       $return['fid'] = $file->fid;
