diff --git a/piwik_stats.info b/piwik_stats.info
index ca3fc96..8c5e3e7 100644
--- a/piwik_stats.info
+++ b/piwik_stats.info
@@ -3,4 +3,4 @@ description = Integrates piwik statistics as entity field.
 package = Statistics
 core = 7.x
 configure = admin/config/system/piwik
-dependencies[] = piwik
+dependencies[] = matomo
diff --git a/piwik_stats.module b/piwik_stats.module
index 51e8610..f24b7cd 100644
--- a/piwik_stats.module
+++ b/piwik_stats.module
@@ -369,11 +369,11 @@ function piwik_stats_flush_caches() {
 /**
  * Implements hook_form_FORM_ID_alter().
  *
- * Form ID: piwik_admin_settings (Settings page of piwik module).
+ * Form ID: matomo_admin_settings (Settings page of piwik module).
  *
  * Adds piwik stats settings to piwik configuration.
  */
-function piwik_stats_form_piwik_admin_settings_form_alter(&$form, &$form_state, $form_id) {
+function piwik_stats_form_matomo_admin_settings_form_alter(&$form, &$form_state, $form_id) {
   $form['account']['piwik_stats_token_auth'] = array(
     '#type' => 'textfield',
     '#title' => t('Piwik authentication token'),
@@ -417,11 +417,11 @@ function piwik_stats_form_piwik_admin_settings_form_alter(&$form, &$form_state,
     '#description' => t('Maximum of time a piwik API request can take.'),
   );
   // Add submit callback to save input.
-  $form['#submit'][] = 'piwik_stats_piwik_admin_settings_form_submit';
+  $form['#submit'][] = 'piwik_stats_matomo_admin_settings_form_submit';
   // Submit button for refreshing piwik_stats table.
   $form['actions']['fill_piwik_stats'] = array(
     '#type' => 'submit',
-    '#submit' => array('piwik_stats_piwik_admin_settings_batch_submit'),
+    '#submit' => array('piwik_stats_matomo_admin_settings_batch_submit'),
     '#value' => t('Fill statistical Piwik fields'),
   );
 }
@@ -429,7 +429,7 @@ function piwik_stats_form_piwik_admin_settings_form_alter(&$form, &$form_state,
 /**
  * Submit callback for piwik configuration form.
  */
-function piwik_stats_piwik_admin_settings_form_submit($form, &$form_state) {
+function piwik_stats_matomo_admin_settings_form_submit($form, &$form_state) {
   // Trim and save authentication token.
   variable_set('piwik_stats_token_auth', trim($form_state['input']['piwik_stats_token_auth']));
 }
@@ -437,7 +437,7 @@ function piwik_stats_piwik_admin_settings_form_submit($form, &$form_state) {
 /**
  * Submit callback for filling piwik_stats fields manually by batch.
  */
-function piwik_stats_piwik_admin_settings_batch_submit($form, &$form_state) {
+function piwik_stats_matomo_admin_settings_batch_submit($form, &$form_state) {
   // We want some fresh data here, remove cached data.
   cache_clear_all('piwik_stats:xml:', 'cache_piwik_stats', TRUE);
 
@@ -622,9 +622,9 @@ function piwik_stats_get_xml_data($field) {
       $result = $cache;
     }
     else {
-      $url = variable_get('piwik_url_http');
+      $url = variable_get('matomo_url_http');
       $token = variable_get('piwik_stats_token_auth');
-      $site_id = variable_get('piwik_site_id');
+      $site_id = variable_get('matomo_site_id');
       // Check if connection credentials are provided.
       if (!isset($url, $token, $site_id)) {
         return FALSE;
@@ -773,8 +773,7 @@ function _piwik_stats_get_aliases($uri) {
       }
     }
   }
-
-  return $urls;
+  return array_unique($urls);
 }
 
 /**
